assignment5/q1.cpp

```diff
@@ -0,0 +1,184 @@
+        letters.insert(word_letter);
+
+        //if word is made only of one character
+        while((letter_id < length - 1) && word_letter == dictionary.at(0)[letter_id]) {
+            letter_id++;
+        }
+        if(letter_id == length - 1) {
+            connections[word_letter] = &(*new vector<char>);
+            (*connections[(int) word_letter]).push_back(word_letter);
+        }
+
+        return;
+    }
+
+    //make connections
+    for(int word_id = 0; word_id < dictionary.size()-1; word_id++) {
+        string word1 = dictionary[word_id];
+        string word2 = dictionary[word_id + 1];
+        int char_id = 0;
+        while(word1[char_id] == word2[char_id])
+            char_id++;
+
+        if(connections[word1[char_id]] == NULL) {
+            letters.insert(word1[char_id]);
+            in_number[word1[char_id]] = 0;
+            connections[word1[char_id]] = &(*new vector<char>);
+        }
+        if(connections[word2[char_id]] == NULL) {
+            letters.insert(word2[char_id]);
+            in_number[word2[char_id]] = 0;
+            connections[word2[char_id]] = &(*new vector<char>);
+        }
+        //add the directed edge
+        (*connections[word1[char_id]]).push_back(word2[char_id]);
+        letters.erase(word2[char_id]);
+        in_number[word2[char_id]]++;
+
+    }
+
+    return;
+
+}
+
+vector<char> make_alphabet_from_rules(vector<char> *connections[256], set<char> &letters, vector<int> &in_number) {
+
+    int current_id = 0;
+
+    vector<char> alphabet;
+    set<char>::iterator current_letter = letters.begin();
+
+    //if dictionary was empty
+    if(letters.empty()) {
+        cout<<"empty"<<endl;
+        return alphabet;
+    }
+
+    //while there are still letters to process
+    while(!letters.empty()) {
+
+        alphabet.push_back(*current_letter);
+
+        //if dictionary had one word consisting of more than one char
+        if(connections[*current_letter] == NULL){
+            cout<<"cannot make rules"<<endl;
+            return alphabet;
+        }
+
+        vector<char> *vector_for_letter = &(*(connections[*current_letter]));
+
+        //look for all connections from letter
+        for(int vertex = 0; vertex < (*vector_for_letter).size(); vertex++) {
+            char vertex_letter = (*vector_for_letter)[vertex];
+            in_number[vertex_letter]--;
+            if(in_number[vertex_letter] == 0) {
+                letters.insert(vertex_letter);
+            }
+        }
+        //delete letter's connections
+        vector<char> *row = &(*(connections[*current_letter]));
+        letters.erase(*current_letter);
+        (*row).erase((*row).begin(), (*row).end());
+        current_letter = letters.begin();
+
+    }
+
+    //if there are still any connections
+    for(int i = 0; i < 256; i++){
+        if((&(*connections[i]) != NULL) && (*connections[i]).empty() != 1) {
+            //if cycle or contradictory rules
+            cout<<"cannot make rules"<<endl;
```

Comment:
This case is really an error. A cycle is an illegal input that needs special care (in case we do not assume legal input which was allowed in this assignment).
There are several correct ways to deal with it. Returning an empty alphabet is ok - you should specify it in the function documentation:
`// Returns an empty alphabet in case no legal ordering is found.`
Throwing an exception is also ok, still should be documented.
Having two return values, `bool` for success and `vector<char>` for alphabet is ok too, though I think it is less cleaner than the solutions above (Can return a pair`<bool, vector<char>>` or return one of them via output parameter).

Either way, try to avoid printing inside the function, and use other ways to signify the error or the bad input.